llvm/unittests/Transforms/Vectorize/SandboxVectorizer/DependencyGraphTest.cpp

```diff
@@ -313,3 +342,278 @@ define void @foo(ptr %ptr, i8 %v0, i8 %v1) {
+
+  auto It = BB->begin();
+  auto *Store0N = DAG.getNode(&*It++);
+  auto *StackSaveN = DAG.getNode(&*It++);
+  auto *StackRestoreN = DAG.getNode(&*It++);
+  auto *Store1N = DAG.getNode(&*It++);
+
+  EXPECT_TRUE(dependency(Store0N, StackSaveN));
+  EXPECT_TRUE(dependency(StackSaveN, StackRestoreN));
+  EXPECT_TRUE(dependency(StackRestoreN, Store1N));
+}
+
+// Make sure there is a dependency between a stackrestore and an alloca.
+TEST_F(DependencyGraphTest, StackRestoreAndInAlloca) {
+  parseIR(C, R"IR(
+declare void @llvm.stackrestore(ptr %ptr)
+
+define void @foo(ptr %ptr) {
+  call void @llvm.stackrestore(ptr %ptr)
+  %alloca0 = alloca inalloca i8              ; Should depend on stackrestore
+  ret void
+}
+)IR");
+  Function *LLVMF = M->getFunction("foo");
+
+  sandboxir::Context Ctx(C);
+  auto *F = Ctx.createFunction(LLVMF);
+  auto *BB = &*F->begin();
+
+  sandboxir::DependencyGraph DAG(getAA(*LLVMF));
+  DAG.extend({&*BB->begin(), BB->getTerminator()->getPrevNode()});
+
+  auto It = BB->begin();
+  auto *StackRestoreN = DAG.getNode(&*It++);
+  auto *AllocaN = DAG.getNode(&*It++);
+
+  EXPECT_TRUE(dependency(StackRestoreN, AllocaN));
+}
+
+// Make sure there is a dependency between the alloca and stacksave
+TEST_F(DependencyGraphTest, StackSaveAndInAlloca) {
+  parseIR(C, R"IR(
+declare ptr @llvm.stacksave()
+
+define void @foo(ptr %ptr) {
+  %alloca0 = alloca inalloca i8              ; Should depend on stackrestore
+  %stack0 = call ptr @llvm.stacksave()       ; Should depend on alloca0
+  ret void
+}
+)IR");
+  Function *LLVMF = M->getFunction("foo");
+
+  sandboxir::Context Ctx(C);
+  auto *F = Ctx.createFunction(LLVMF);
+  auto *BB = &*F->begin();
+
+  sandboxir::DependencyGraph DAG(getAA(*LLVMF));
+  DAG.extend({&*BB->begin(), BB->getTerminator()->getPrevNode()});
+
+  auto It = BB->begin();
+  auto *AllocaN = DAG.getNode(&*It++);
+  auto *StackSaveN = DAG.getNode(&*It++);
+
+  EXPECT_TRUE(dependency(AllocaN, StackSaveN));
+}
+
+// A non-InAlloca in a stacksave-stackrestore region does not need extra
+// dependencies.
+TEST_F(DependencyGraphTest, StackSaveRestoreNoInAlloca) {
+  parseIR(C, R"IR(
+declare ptr @llvm.stacksave()
+declare void @llvm.stackrestore(ptr %ptr)
+declare void @use(ptr %ptr)
+
+define void @foo() {
+  %stack = call ptr @llvm.stacksave()
+  %alloca1 = alloca i8                         ; No dependency
+  call void @llvm.stackrestore(ptr %stack)
+  ret void
+}
+)IR");
+  Function *LLVMF = M->getFunction("foo");
+
+  sandboxir::Context Ctx(C);
+  auto *F = Ctx.createFunction(LLVMF);
+  auto *BB = &*F->begin();
+
+  sandboxir::DependencyGraph DAG(getAA(*LLVMF));
+  DAG.extend({&*BB->begin(), BB->getTerminator()->getPrevNode()});
+
+  auto It = BB->begin();
+  auto *StackSaveN = DAG.getNode(&*It++);
+  auto *AllocaN = DAG.getNode(&*It++);
+  auto *StackRestoreN = DAG.getNode(&*It++);
+
+  EXPECT_FALSE(dependency(StackSaveN, AllocaN));
+  EXPECT_FALSE(dependency(AllocaN, StackRestoreN));
+}
```

Comment:
How about checking dependency types?

Reply:
The dependency type is not visible externally, these are all private members. We don't really need to know the exact dependency ype for our use case.